src/modules/ztsi/src/lib/Ztsi.cpp

```diff
@@ -158,201 +160,329 @@ int Ztsi::SetEnabled(bool enabled)
+    if ((0 == status) || (EINVAL == status))
     {
-        // Check if the serviceUrl is already set to the desired value
-        if (serviceUrl != config.serviceUrl)
+        if (serviceUrl != configuration.serviceUrl)
         {
-            config.serviceUrl = serviceUrl;
-            status = WriteAgentConfig(config);
+            configuration.serviceUrl = serviceUrl;
+            status = WriteAgentConfiguration(configuration);
         }
     }
     else if (ENOENT == status)
     {
-        // If the config file doesn't exist, create it with the desired serviceUrl
-        config.enabled = g_defaultEnabledState;
-        config.serviceUrl = serviceUrl;
-        status = WriteAgentConfig(config);
-    }
-    else
-    {
-        OsConfigLogError(ZtsiLog::Get(), "Failed to set serviceUrl property");
+        // If the configuration file doesn't exist, create it with the desired serviceUrl
+        configuration.enabled = g_defaultEnabledState;
+        configuration.serviceUrl = serviceUrl;
+        status = CreateConfigurationFile(configuration);
     }
 
     return status;
 }
 
-bool Ztsi::IsValidConfig(const Ztsi::AgentConfig& config)
+bool Ztsi::IsValidConfiguration(const Ztsi::AgentConfiguration& configuration)
 {
     bool isValid = true;
 
-    if (config.serviceUrl.empty() && config.enabled)
+    if (configuration.serviceUrl.empty() && configuration.enabled)
     {
         OsConfigLogError(ZtsiLog::Get(), "ServiceUrl is empty and enabled is true");
         isValid = false;
     }
 
     std::regex urlPattern(g_urlRegex);
-    if (!config.serviceUrl.empty() && !regex_match(config.serviceUrl, urlPattern))
+    if (!configuration.serviceUrl.empty() && !regex_match(configuration.serviceUrl, urlPattern))
     {
-        OsConfigLogError(ZtsiLog::Get(), "Invalid serviceUrl '%s'", config.serviceUrl.c_str());
+        OsConfigLogError(ZtsiLog::Get(), "Invalid serviceUrl '%s'", configuration.serviceUrl.c_str());
         isValid = false;
     }
 
     return isValid;
 }
 
-int Ztsi::ReadAgentConfig(Ztsi::AgentConfig& config)
+bool Ztsi::FileExists(const std::string& filePath)
 {
-    int status = 0;
-
-    std::ifstream file(m_agentConfigFile);
-    if (!file.good())
-    {
-        status = ENOENT;
-        return status;
-    }
+    struct stat sb;
+    return (0 == stat(filePath.c_str(), &sb) && S_ISREG(sb.st_mode));
+}
 
-    rapidjson::IStreamWrapper isw(file);
-    rapidjson::Document document;
-    if (document.ParseStream(isw).HasParseError())
-    {
-        OsConfigLogError(ZtsiLog::Get(), "Failed to parse config file: %s", m_agentConfigFile.c_str());
-        status = EINVAL;
-        return status;
-    }
+std::FILE* Ztsi::OpenAndLockFile(const char* mode)
+{
+    int fd = -1;
+    std::FILE* fp = nullptr;
 
-    if (document.HasMember(g_configPropertyEnabled))
+    if (FileExists(m_agentConfigurationFile) && (nullptr != (fp = fopen(m_agentConfigurationFile.c_str(), mode))))
```

Comment:
Isn't this subject to race conditions? What if the file gets deleted right after the `FileExists()` but before the `fopen()` call? Isn't the `FileExists()` call simply redundant anyway?

Reply:
`FileExists()` is there to ensure we only open the file if it has been created. Now looking at it I see that `FileExists()` can be removed there since we only ever open the file using a mode that will not also create a file if it does not exist (`r` `r+`).

Reply:
Correct, I can remove this `FileExists()` in `OpenAndLock()`